common/types_spec.go

```diff
@@ -226,3 +221,129 @@ func (r *BuilderBlockValidationRequest) MarshalJSON() ([]byte, error) {
+	case consensusspec.DataVersionDeneb:
+		return json.Marshal(r.Deneb)
+	case consensusspec.DataVersionUnknown, consensusspec.DataVersionPhase0, consensusspec.DataVersionAltair, consensusspec.DataVersionBellatrix:
+		fallthrough
+	default:
+		return nil, errors.Wrap(ErrInvalidVersion, fmt.Sprintf("%d is not supported", r.Version))
+	}
+}
+
+func (r *VersionedSubmitBlockRequest) UnmarshalJSON(input []byte) error {
+	var err error
+
+	deneb := new(deneb.SubmitBlockRequest)
+	if err = json.Unmarshal(input, deneb); err == nil {
+		r.Version = consensusspec.DataVersionDeneb
+		r.Deneb = deneb
+		return nil
+	}
+	capella := new(capella.SubmitBlockRequest)
+	if err = json.Unmarshal(input, capella); err == nil {
+		r.Version = consensusspec.DataVersionCapella
+		r.Capella = capella
+		return nil
+	}
+	return errors.Wrap(err, "failed to unmarshal SubmitBlockRequest")
+}
+
+type VersionedSignedBlockRequest struct {
+	consensusapi.VersionedBlockRequest
+}
+
+func (r *VersionedSignedBlockRequest) MarshalJSON() ([]byte, error) {
+	switch r.Version {
+	case consensusspec.DataVersionCapella:
+		return json.Marshal(r.Capella)
+	case consensusspec.DataVersionDeneb:
+		return json.Marshal(r.Deneb)
+	case consensusspec.DataVersionUnknown, consensusspec.DataVersionPhase0, consensusspec.DataVersionAltair, consensusspec.DataVersionBellatrix:
+		fallthrough
+	default:
+		return nil, errors.Wrap(ErrInvalidVersion, fmt.Sprintf("%d is not supported", r.Version))
+	}
+}
+
+func (r *VersionedSignedBlockRequest) UnmarshalJSON(input []byte) error {
+	var err error
+
+	deneb := new(apiv1deneb.SignedBlockContents)
+	if err = json.Unmarshal(input, deneb); err == nil {
+		r.Version = consensusspec.DataVersionDeneb
+		r.Deneb = deneb
+		return nil
+	}
+
+	capella := new(consensuscapella.SignedBeaconBlock)
+	if err = json.Unmarshal(input, capella); err == nil {
+		r.Version = consensusspec.DataVersionCapella
+		r.Capella = capella
+		return nil
+	}
+	return errors.Wrap(err, "failed to unmarshal SignedBeaconBlockRequest")
+}
+
+type VersionedSignedBlindedBlockRequest struct {
+	consensusapi.VersionedBlindedBlockRequest
+}
+
+func (r *VersionedSignedBlindedBlockRequest) MarshalJSON() ([]byte, error) {
+	switch r.Version {
+	case consensusspec.DataVersionCapella:
+		return json.Marshal(r.Capella)
+	case consensusspec.DataVersionDeneb:
+		return json.Marshal(r.Deneb)
+	case consensusspec.DataVersionUnknown, consensusspec.DataVersionPhase0, consensusspec.DataVersionAltair, consensusspec.DataVersionBellatrix:
+		fallthrough
+	default:
+		return nil, errors.Wrap(ErrInvalidVersion, fmt.Sprintf("%d is not supported", r.Version))
+	}
+}
+
+func (r *VersionedSignedBlindedBlockRequest) UnmarshalJSON(input []byte) error {
+	var err error
+
+	deneb := new(apiv1deneb.SignedBlindedBlockContents)
+	if err = json.Unmarshal(input, deneb); err == nil {
+		r.Version = consensusspec.DataVersionDeneb
+		r.Deneb = deneb
+		return nil
+	}
+
+	capella := new(apiv1capella.SignedBlindedBeaconBlock)
```

Comment:
Nit: these variables (`deneb` and `capella`) collide with package names.

I think we should fix this by renaming the packages (in a later PR).